### ISSUE TITLE
Investigate consolidating API requests for the detail page into a single API call

### ISSUE THREAD

**Issue by diox** [mozilla/addons-frontend#10038]
Like https://github.com/mozilla/addons-frontend/issues/9996 but for the detail page.

Investigate if that really makes sense - we may not want to have a single endpoint, but having a fast first render, with all the base info needed. So for instance, maybe we could get rid of the call to the ratings API is we rolled `grouped_ratings` into the detail API endpoint ? But that would only work if the frontend is able to leverage this in an efficient way.

**Comment by bobsilverberg** [comment]
There are still a couple of potential areas for improvement even after #10039 is fixed:

1. The call to `/ratings/rating/?addon=123&show_grouped_ratings=true` retrieves not just the grouped ratings, but also the first page of actual ratings, which is not needed for the detail page, so ideally we would have a way of retrieving just the grouped ratings without any detail.
2. We issue a second request to the `ratings` API on the detail page: `/ratings/rating/?user=789&addon=123`. This is done to retrieve the rating from the current user. I wonder if we could update the endpoint to allow for retrieval of both the grouped ratings and the user's current rating (and exclude the detail as discussed in 1, above), or perhaps we consider adding both of those pieces of data to the add-on detail endpoint. If the user is not logged in, we don't try to retrieve their rating, but that could easily be dealt with by simply not passing a `user` as a param.

**Comment by willdurand** [comment]
> The call to `/ratings/rating/?addon=123&show_grouped_ratings=true` retrieves not just the grouped ratings, but also the first page of actual ratings, which is not needed for the detail page, so ideally we would have a way of retrieving just the grouped ratings without any detail.

it'd be interesting to try to understand why we're retrieving the first page of ratings.

**Comment by bobsilverberg** [comment]
> > The call to `/ratings/rating/?addon=123&show_grouped_ratings=true` retrieves not just the grouped ratings, but also the first page of actual ratings, which is not needed for the detail page, so ideally we would have a way of retrieving just the grouped ratings without any detail.
> 
> it'd be interesting to try to understand why we're retrieving the first page of ratings.

I assume it's because the endpoint doesn't specifically provide a way to exclude the ratings results, although we might be able to hack around that with `pagesize=0`. I'm not sure that anyone has ever taken a look at this and realized we were retrieving results when we didn't need them. 

At the very least I think we should update the endpoint to allow for retrieving only grouped ratings (or provide a different endpoint for that), but I think it might be worth going further and providing an endpoint that can be used for both grouped ratings and the user's current rating, in a single request, or perhaps going even further and adding those pieces of data to the add-on detail response (either selectable via a param, or perhaps a unique add-on detail endpoint designed for AMO's add-on detail page).

**Comment by bobsilverberg** [comment]
For reference, here's a list of the APIs called for the add-on detail page:
- https://addons.mozilla.org/api/v5/addons/addon/{slug}/
	- This gets the detail for the add-on.

- https://addons.mozilla.org/api/v5/ratings/rating/?addon={addonId}&show_grouped_ratings=true
	- As discussed above, this gets the grouped ratings for the Ratings widget

- https://addons.mozilla.org/api/v5/addons/search/?app=firefox&appversion=85.0&author={userId}&exclude_addons={slug}&page=1&page_size=6&sort=hotness&type=extension
	- This gets the list of other add-ons by the same author, used by the "More extensions by {Author Name}" widget. Note that this is not cached in the redux store, so if another detail page for the same author is visited, we call the API again. This is likely by design as that's not something that would be particularly common.

- https://addons.mozilla.org/api/v5/addons/recommendations/?guid={guid}&recommended=true
	- This gets the recommendations for the "Recommended add-ons widget"

- https://addons.mozilla.org/api/v5/ratings/rating/?user={userId}&addon={addonId}
	- This gets the current user's rating for the add-on. Note that this is only called if the user is logged in.

- https://addons.mozilla.org/api/v5/accounts/account/{userId}/collections/
	- This gets the list of collections for the user, used by the "Add to Collection" widget. Note that this is only called the first time it is needed, and the data is cached in the redux store. Note also that this is only called if the user is logged in.

**Comment by bobsilverberg** [comment]
Based on the above, here are some thoughts:

1. If we can get the current user's rating, and also the grouped ratings, as part of the call to the add-on detail API, that would save us two API calls. This data is small, so shouldn't greatly impact the amount of data returned by the detail API. We could have a flag, or flags to include it, or perhaps we could create a new endpoint designed specifically for the add-on detail page.

2. We always call the APIs to retrieve the recommendations and the other extensions by the author, so we could consider consolidating those responses as well into an endpoint that is designed to serve the add-on detail page.

3. I'm not too worried about the collections API call. We're going to have to retrieve those once for the user. We could consider doing that when a user logs in, or first accesses the site from anywhere, as opposed to only when we need it, but that's not going to have any impact on loading the add-on detail page as the first page loaded. 

What are people's thoughts on the above?

**Comment by diox** [comment]
1. I think it makes sense to have some kind of parameter to include that data. Fetching grouped ratings can be costly, so it shouldn't be done by default - we might want to take a look at the query to see if we can speed it up while we're at it.
2. Maybe. Both are always called client-side, right ? The thing that make me hesitate about grouping those together is that one is super slow because it depends on TAAR and not the other. So maybe there is more investigation to be done in addons-server first.
3. Not too worried about that for the time being either.

**Comment by willdurand** [comment]
>     2\. We always call the APIs to retrieve the recommendations and the other extensions by the author, so we could consider consolidating those responses as well into an endpoint that is designed to serve the add-on detail page.

I don't think we pass the client ID to the frontend's server code so the recommendation API should be kept separate I believe.

**Comment by eviljeff** [comment]
Filed mozilla/addons-server#16457 for the API work for the ratings

**Comment by bobsilverberg** [comment]
Just coming back to this issue to see if there is still work to be done here. Looking at my suggestions from https://github.com/mozilla/addons-frontend/issues/10038#issuecomment-771815288:

1. We are now getting the grouped ratings along with the add-on detail, so that's done. There is still the possibility of retrieving the current user's rating at the same time which should be discussed. I'll open a separate issue for that.
2. It seems like there is consensus that we should keep the API calls for recommendations and for "other add-ons by this author" as they are, i.e., as separate API calls, so we can consider this done (via wontfix).
3. I think there is also consensus that the call to the collections API issue is a wontfix.

Based on the above, it looks like we've done everything we think we can do, regarding API calls made from the detail page, with the possible exception of the user's rating, for which I am going to open a separate issue. Therefore I think this issue can be closed.

Does anyone disagree with that assessment?

**Comment by eviljeff** [comment]
> Based on the above, it looks like we've done everything we think we can do, regarding API calls made from the detail page, with the possible exception of the user's rating, for which I am going to open a separate issue. 

I'm not sure we need to do anything about separate calls for the user's rating - it only affects logged in users (we bypass a whole bunch of optimization when the user is authenticated so it's far from the only bottleneck for that use case).

**Comment by bobsilverberg** [comment]
> > Based on the above, it looks like we've done everything we think we can do, regarding API calls made from the detail page, with the possible exception of the user's rating, for which I am going to open a separate issue.
> 
> I'm not sure we need to do anything about separate calls for the user's rating - it only affects logged in users (we bypass a whole bunch of optimization when the user is authenticated so it's far from the only bottleneck for that use case).

Yes, and I seem to recall that we discussed this before, but I cannot find a record of that discussion. There were other issues with the way we store reviews that would make it quite complicated to implement if we wanted to get the data from the add-on detail endpoint. So I don't think I need to open a separate issue about it. I believe there is agreement that we will leave this API request as is as well.
